returnn/tf/util/data.py

```diff
@@ -2337,6 +2337,32 @@ def copy_extend_batch(self, batch):
         data.placeholder = x
     return data
 
+  def copy_unpack_batch(self):
+    """
+    If the batch dim contains a :class:`BatchInfo.PackedDim`, unpack it and restore the data from before the packing.
+
+    :rtype: Data
+    """
+    assert self.have_batch_axis()
+    assert self.batch, "%s: batch unset" % self
+    data = self.copy()
+    kwargs = self.get_kwargs(include_special_axes=False)
+
+    dim_tags = []
+    for dim_tag in data.dim_tags:
+      if dim_tag.is_batch_dim() and hasattr(dim_tag.batch, "virtual_dims"):
```

Comment:
You should not use `hasattr`. Why is that needed?
Maybe you had some cases where `dim_tag.batch` was `None`. Then check for that.
```suggestion
      if dim_tag.is_batch_dim() and dim_tag.batch:
```

Reply:
I don't think I had cases with `dim_tag.batch=None` but was thinking of the case that there are no virtual dims, but of course the attribute then exists anyway and contains an empty list.

Reply:
The batch usually always has virtual dims, specifically at least the `GlobalBatchDim`. (Although no virtual dims at all are allowed, and means that this is a batch with dim 1 then, which is useful for broadcasting.)

Reply:
> I don't think I had cases with `dim_tag.batch=None`

Ok, then don't include that in the check. It should be invalid. You can add `assert dim_tag.batch` if you want.